### ISSUE TITLE
False positive in recursive type-checking

### ISSUE THREAD

**Issue by arnimarj** [python/mypy#21466]
**Bug Report**

When experimenting with the recently introduced `JsonType` in the `requests` library I came across a seeming false-positive on my codebase:

```python
from typing import TypeAlias

JsonType: TypeAlias = (
    bool | None | int | bool | float | str | list["JsonType"] | dict[str, "JsonType"]
)


str_bool: dict[str, bool] = {}

# mypy gives error for this line
a: JsonType = {'attributes': str_bool}
# but not this
b: JsonType = {'attributes': {'str': True}}
```

**Expected Behavior**

The value assigned to `a` seems to satisfy `JsonType` and the given error is unexpected.

```shell
# mypy --strict recursion.py
recursion.py:10: error: Dict entry 0 has incompatible type "str": "dict[str, bool]"; expected "str": "int | float | str | list[JsonType] | dict[str, JsonType] | None"  [dict-item]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

Tested on mypy 2.1 with the `--strict` flag.

**Comment by hauntsaninja** [comment]
This is expected, because dict is invariant in its value type. 

```
str_bool: dict[str, bool] = {}
def oops(a: JsonType):
    if isinstance(a["attributes"], dict):
        a["attributes"]["val"] = 123  # valid json

oops({'attributes': str_bool})

assert all(isinstance(v, bool) for v in str_bool.values())  # oops
```

**Comment by hauntsaninja** [comment]
See also https://github.com/python/typing/issues/182#issuecomment-1320974824
